google/cloud/aiplatform/metadata/experiment_run_resource.py

```diff
@@ -0,0 +1,1184 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import concurrent.futures
+import functools
+from typing import Callable, Dict, List, Optional, Set, Union, Any
+
+from google.api_core import exceptions
+from google.auth import credentials as auth_credentials
+from google.protobuf import timestamp_pb2
+
+import google.cloud.aiplatform.metadata.constants
+from google.cloud.aiplatform import base
+from google.cloud.aiplatform import initializer
+from google.cloud.aiplatform import pipeline_jobs
+from google.cloud.aiplatform.compat.types import execution as gca_execution
+from google.cloud.aiplatform.compat.types import (
+    tensorboard_time_series as gca_tensorboard_time_series,
+)
+from google.cloud.aiplatform.metadata import artifact
+from google.cloud.aiplatform.metadata import constants
+from google.cloud.aiplatform.metadata import context
+from google.cloud.aiplatform.metadata import execution
+from google.cloud.aiplatform.metadata import experiment_resources
+from google.cloud.aiplatform.metadata import metadata
+from google.cloud.aiplatform.metadata import resource
+from google.cloud.aiplatform.metadata import utils as metadata_utils
+from google.cloud.aiplatform.tensorboard import tensorboard_resource
+
+
+_LOGGER = base.Logger(__name__)
+
+
+def _format_experiment_run_resource_id(experiment_name: str, run_name: str) -> str:
+    """Formats the the experiment run resource id as a concatenation of experiment name and run name.
+
+    Args:
+        experiment_name (str): Name of the experiment which is it's resource id.
+        run_name (str): Name of the run.
+    Returns:
+        The resource id to be used with this run.
+    """
+    return f"{experiment_name}-{run_name}"
+
+
+def _v1_not_supported(method: Callable) -> Callable:
+    """Helpers wrapper for backward compatibility. Raises when using an API not support for legacy runs."""
+
+    @functools.wraps(method)
+    def wrapper(self, *args, **kwargs):
+        if isinstance(self._metadata_node, execution.Execution):
+            raise NotImplementedError(
+                f"{self._run_name} is an Execution run created during Vertex Experiment Preview and does not support"
+                f" {method.__name__}. Please create a new Experiment run to use this method."
```

Comment:
Are there any reasonably-common scenarios where a user might want to go back and re-run old runs?

I don't have a problem with this wrapper, just trying to estimate how annoying the lack of backwards compatibility might be.

Reply:
This wrapper doesn't stop the user from re-running old runs. It prevents using the newly added APIs on old runs because of the difference in the data model. They can still use APIs that were available on the run in Preview.